src/algebra/direct_sum_graded.lean

```diff
@@ -251,105 +251,86 @@ end one
 section mul
 variables [has_add ι] [Π i, add_comm_monoid (A i)] [ghas_mul A]
 
+open add_monoid_hom (map_zero map_add flip_apply coe_comp comp_hom_apply_apply)
+
+/-- The multiplication from the `has_mul` instance, as a bundled homomorphism. -/
+def mul_hom : (⨁ i, A i) →+ (⨁ i, A i) →+ ⨁ i, A i :=
+direct_sum.to_add_monoid $ λ i,
+  add_monoid_hom.flip $ direct_sum.to_add_monoid $ λ j, add_monoid_hom.flip $
+    (direct_sum.of A _).comp_hom.comp ghas_mul.mul
+
 instance : has_mul (⨁ i, A i) :=
-{ mul := λ a b, begin
-    -- the elaborator struggles here, so use tactics to assemble the expression
-    refine direct_sum.to_add_monoid (λ j,
-      direct_sum.to_add_monoid (λ i,
-        (add_monoid_hom.comp_hom (direct_sum.of A $ i + j)).comp _
-      ) a
-    ) b,
-    exact ghas_mul.mul,
-  end }
+{ mul := λ a b, mul_hom A a b }
 
 instance : mul_zero_class (⨁ i, A i) :=
 { mul := (*),
   zero := 0,
-  zero_mul := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of] },
-  mul_zero := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of] } }
+  zero_mul := λ a, by { unfold has_mul.mul, simp only [map_zero, add_monoid_hom.zero_apply]},
+  mul_zero := λ a, by { unfold has_mul.mul, simp only [map_zero] } }
 
 instance : distrib (⨁ i, A i) :=
 { mul := (*),
   add := (+),
-  left_distrib := by { unfold has_mul.mul, simp, },
-  right_distrib := by { unfold has_mul.mul, simp [direct_sum.to_add_monoid, direct_sum.of], } }
+  left_distrib := λ a b c, by { unfold has_mul.mul, simp only [map_add]},
+  right_distrib := λ a b c, by { unfold has_mul.mul, simp only [map_add, add_monoid_hom.add_apply]}}
 
-lemma of_mul_of {i j} (a : A i) (b : A j) :
-  of _ i a * of _ j b = of _ (i + j) (ghas_mul.mul a b) :=
+variables {A}
+
+lemma of_mul_of' {i j} (a : A i) (b : A j) :
+  mul_hom A (of _ i a) (of _ j b) = of _ (i + j) (ghas_mul.mul a b) :=
 begin
-  unfold has_one.one has_mul.mul,
-  simp,
+  unfold mul_hom,
+  rw [to_add_monoid_of, flip_apply, to_add_monoid_of, flip_apply, coe_comp, function.comp_app,
+      comp_hom_apply_apply, coe_comp, function.comp_app],
 end
 
+lemma of_mul_of {i j} (a : A i) (b : A j) :
+  of _ i a * of _ j b = of _ (i + j) (ghas_mul.mul a b) :=
+of_mul_of' a b
+
 end mul
 
-/-! The various semiring fields are proved separately because the proofs are slow. -/
+-- Like `add_monoid_hom.congr_fun`, but convenient for a simple argument and complicated equality.
+-- A non-private version might be called `add_monoid_hom.congr_at`.
```

Comment:
Instead of defining `congr_at`, does `flip add_monoid_hom.congr_fun` work? Adding this definition to every bundled hom (as public API) just to swap the argument order feels like a poor tradeoff.

Alternatively, could you just use `refine congr_fun _ x,` in a tactic block, or is that harder than I'm anticipating?

Reply:
Yeah, I made this a private abbreviation because it definitely felt like it'd be a bit much to add it to every bundled hom.

Playing around some more, I can do it with a `suffices`:
```
private lemma mul_assoc (a b c : ⨁ i, A i) : a * b * c = a * (b * c) :=
suffices (mul_hom A).comp_hom.comp (mul_hom A)           -- `λ a b c, a * b * c` as a bundled hom
   = (add_monoid_hom.comp_hom flip_hom $
       (mul_hom A).flip.comp_hom.comp (mul_hom A)).flip, -- `λ a b c, a * (b * c)` as a bundled hom
  from add_monoid_hom.congr_fun (add_monoid_hom.congr_fun (add_monoid_hom.congr_fun this a) b) c,
begin
```
I think the `suffices` actually puts this logic in a slightly cleaner order to read, too. I'll revise them all to use this arrangement, and drop the abbreviation.

---

Other things I tried just now:

`flip add_monoid_hom.congr_fun` doesn't work, though I don't totally understand why:
```
type mismatch at application
  flip add_monoid_hom.congr_fun
term
  add_monoid_hom.congr_fun
has type
  ?m_5 = ?m_6 → ∀ (x : ?m_1), ⇑?m_5 x = ⇑?m_6 x : Prop
but is expected to have type
  ?m_1 → ?m_2 → ?m_3 : Sort (imax ? ? ?)
```
That's from `#check flip add_monoid_hom.congr_fun`; I get similar errors, just with some metavariables filled in, if I try to use it in place of `congr_at`.

I tried `refine`, but the elaborator doesn't like that -- I think (perhaps understandably) it wants to be told what the type of that hypothesis should be:
```
invalid type ascription, term has type
  ⇑?m_3 x = ⇑?m_4 x
but is expected to have type
  1 * x = x
```
For `one_mul` we can just put there the type the current version puts in a `show`:
```
  refine add_monoid_hom.congr_fun (_ : mul_hom A 1 = add_monoid_hom.id (⨁ i, A i)) x,
```
But for `mul_assoc` I think that would mean exhibiting not only the pair of bundled homs on three arguments a, b, c (for the last `refine`) but also the pair of those with a fixed (for the one before that) and those with a and b fixed (for the first one.) Whereas with the term-mode version, I guess it's able to start from the fully-homified equality that's given by `show` as the type of the inner term, and compute the intermediate versions as it works outward from there.

Reply:
The `suffices` approach is nice and clear, thanks!